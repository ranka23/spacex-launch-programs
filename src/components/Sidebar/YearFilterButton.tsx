import React from "react";
import { RootStateOrAny, useDispatch, useSelector } from "react-redux";
import * as filterActions from "../../store/filterReducer";
import createYearArray from "../../utils/createYearArray";
import Button from "../UI/Button";
import styles from "./styles.module.scss";

interface ToggleButtonProps {
  filterTitle: string;
}

const YearFilterButton: React.FC<ToggleButtonProps> = ({
  filterTitle,
}: ToggleButtonProps) => {
  const dispatch = useDispatch();

  const yearButton = useSelector(
    (state: RootStateOrAny) => state.filterButton.yearValue
  );

  const yearArray = createYearArray();

  const yearButtonClicked = (value: string) => {
    const yearBtn = yearButton;
    if (value === yearBtn) {
      value = null;
    }

    dispatch(filterActions.setYearValue(value));
  };

  return (
    <div className={styles.filter_container}>
      <h5>{filterTitle}</h5>
      <div className={styles.toggle_button_container}>
        {yearArray.map((year) => {
          return (
            <Button
              className={
                yearButton && yearButton === year ? styles.button_active : null
              }
              key={year}
              toggleValue={year}
              onClick={() => yearButtonClicked(year)}
            />
          );
        })}
      </div>
    </div>
  );
};

export default YearFilterButton;
