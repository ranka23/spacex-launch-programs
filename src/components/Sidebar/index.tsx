import React from "react";
import { RootStateOrAny, useDispatch, useSelector } from "react-redux";
import * as filterActions from "../../store/filterReducer";
import Button from "../UI/Button";
import styles from "./styles.module.scss";
import YearFilterButton from "./YearFilterButton";

const Sidebar: React.FC = () => {
  const dispatch = useDispatch();

  const launchButton = useSelector(
    (state: RootStateOrAny) => state.filterButton.launchValue
  );

  const landingButton = useSelector(
    (state: RootStateOrAny) => state.filterButton.landingValue
  );

  const launchButtonClicked = (value: boolean) => {
    const launchBtn = launchButton;
    if (value === launchBtn) {
      value = null;
    }

    dispatch(filterActions.setLaunchValue(value));
  };

  const landingButtonClicked = (value: boolean) => {
    const landingBtn = landingButton;
    if (value === landingBtn) {
      value = null;
    }

    dispatch(filterActions.setLandingValue(value));
  };

  return (
    <div className={styles.sidebar_container}>
      <h4 className={styles.filter_title}>Filters</h4>
      <YearFilterButton filterTitle="Launch Year" />
      <div className={styles.filter_container}>
        <h5>Successful Launch</h5>
        <div className={styles.toggle_button_container}>
          <Button
            className={
              launchButton && launchButton !== null
                ? styles.button_active
                : null
            }
            onClick={() => launchButtonClicked(true)}
            toggleValue
          />
          <Button
            className={
              !launchButton && launchButton !== null
                ? styles.button_active
                : null
            }
            onClick={() => launchButtonClicked(false)}
            toggleValue={false}
          />
        </div>
      </div>
      <h5>Successful Landing</h5>
      <div className={styles.toggle_button_container}>
        <Button
          className={
            landingButton && landingButton !== null
              ? styles.button_active
              : null
          }
          onClick={() => landingButtonClicked(true)}
          toggleValue
        />
        <Button
          className={
            !landingButton && landingButton !== null
              ? styles.button_active
              : null
          }
          onClick={() => landingButtonClicked(false)}
          toggleValue={false}
        />
      </div>
    </div>
  );
};

export default Sidebar;
