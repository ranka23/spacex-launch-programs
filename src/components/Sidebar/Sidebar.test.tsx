import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import mockStore from "../../utils/mockStore";
import Sidebar from "./index";

describe("<Sidebar />", () => {
  const renderHelper = (reducer = { readyStatus: "invalid" }) => {
    const { dispatch, ProviderWithStore } = mockStore({ spaceXData: reducer });
    const { container } = render(
      <ProviderWithStore>
        <MemoryRouter>
          <Sidebar />
        </MemoryRouter>
      </ProviderWithStore>
    );

    return { dispatch, firstChild: container.firstChild };
  };

  it("Should fetch launchValue when page loads", () => {
    const { dispatch } = renderHelper();

    expect(dispatch).toHaveBeenCalledTimes(1);
  });
});
