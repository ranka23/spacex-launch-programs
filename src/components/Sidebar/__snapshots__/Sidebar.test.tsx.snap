// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> Renders the component 1`] = `
<div
  class="sidebar_container"
>
  <h4
    class="filter_title"
  >
    Filters
  </h4>
  <div
    class="filter_container"
  >
    <h5>
      Launch Year
    </h5>
    <div
      class="toggle_button_container"
    >
      <button
        class="button null"
        type="button"
      >
        2006
      </button>
      <button
        class="button null"
        type="button"
      >
        2007
      </button>
      <button
        class="button null"
        type="button"
      >
        2008
      </button>
      <button
        class="button null"
        type="button"
      >
        2009
      </button>
      <button
        class="button null"
        type="button"
      >
        2010
      </button>
      <button
        class="button null"
        type="button"
      >
        2011
      </button>
      <button
        class="button null"
        type="button"
      >
        2012
      </button>
      <button
        class="button null"
        type="button"
      >
        2013
      </button>
      <button
        class="button null"
        type="button"
      >
        2014
      </button>
      <button
        class="button null"
        type="button"
      >
        2015
      </button>
      <button
        class="button null"
        type="button"
      >
        2016
      </button>
      <button
        class="button null"
        type="button"
      >
        2017
      </button>
      <button
        class="button null"
        type="button"
      >
        2018
      </button>
      <button
        class="button null"
        type="button"
      >
        2019
      </button>
      <button
        class="button null"
        type="button"
      >
        2020
      </button>
    </div>
  </div>
  <div
    class="filter_container"
  >
    <h5>
      Successful Launch
    </h5>
    <div
      class="toggle_button_container"
    >
      <button
        class="button null"
        type="button"
      >
        true
      </button>
      <button
        class="button button_active"
        type="button"
      >
        false
      </button>
    </div>
  </div>
  <h5>
    Successful Landing
  </h5>
  <div
    class="toggle_button_container"
  >
    <button
      class="button button_active"
      type="button"
    >
      true
    </button>
    <button
      class="button null"
      type="button"
    >
      false
    </button>
  </div>
</div>
`;
